sdk/cosmos/microsoft-azure-cosmos/src/main/java/com/azure/data/cosmos/internal/changefeed/implementation/LeaseRenewerImpl.java

```diff
@@ -34,66 +34,76 @@ public LeaseRenewerImpl(Lease lease, LeaseManager leaseManager, Duration leaseRe
-                            Thread.sleep(100);
-                            remainingWork -= 100;
-                        }
-                    } catch (InterruptedException ex) {
-                        // exception caught
-                        logger.info(String.format("Partition %s: renewer task stopped.", self.lease.getLeaseToken()));
-                        break;
-                    }
-                }
-            } catch (RuntimeException ex) {
-                logger.error(String.format("Partition %s: renew lease loop failed.", self.lease.getLeaseToken()), ex);
-                self.resultException = ex;
-            }
-        });
+                return !cancellationToken.isCancellationRequested();
+            })
+            .then()
+            .onErrorResume(throwable -> {
+                logger.error(String.format("Partition %s: renew lease loop failed.", self.lease.getLeaseToken()), throwable);
+                return Mono.error(throwable);
+            });
     }
 
     @Override
     public RuntimeException getResultException() {
         return this.resultException;
     }
 
-    private Mono<Void> renew() {
+    private Mono<Lease> renew(CancellationToken cancellationToken) {
         LeaseRenewerImpl self = this;
 
-        return Mono.fromRunnable( () -> {
-            try {
-                Lease renewedLease = self.leaseManager.renew(this.lease).block();
-                if (renewedLease != null) this.lease = renewedLease;
+        if (cancellationToken.isCancellationRequested()) return Mono.empty();
 
+        return self.leaseManager.renew(self.lease)
+            .map(renewedLease -> {
+                if (renewedLease != null) {
+                    self.lease = renewedLease;
+                }
                 logger.info(String.format("Partition %s: renewed lease with result %s", self.lease.getLeaseToken(), renewedLease != null));
-            } catch (LeaseLostException leaseLostException) {
-                logger.error(String.format("Partition %s: lost lease on renew.", self.lease.getLeaseToken()), leaseLostException);
-                self.resultException = leaseLostException;
-                throw leaseLostException;
-            } catch (Exception ex) {
-                logger.error(String.format("Partition %s: failed to renew lease.", self.lease.getLeaseToken()), ex);
-            }
-        });
+                return renewedLease;
+            })
+            .onErrorResume(throwable -> {
+                if (throwable instanceof LeaseLostException) {
+                    LeaseLostException lle = (LeaseLostException) throwable;
+                    self.resultException = lle;
+                    logger.error(String.format("Partition %s: lost lease on renew.", self.lease.getLeaseToken()), lle);
+                    return Mono.error(lle);
+                }
+
+                logger.error(String.format("Partition %s: failed to renew lease.", self.lease.getLeaseToken()), throwable);
```

Comment:
last comment on the String.format inside slf4j logger.

slf4j logger has support for place holder {}, you don't need String.format here. remove String.format
https://www.slf4j.org/manual.html

please fix this here and elsewhere throughout the code.

Reply:
Fixed.